lib/rules/arrow-body-style.js

```diff
@@ -99,6 +99,24 @@ module.exports = {
             return sourceCode.getTokenAfter(node);
         }
 
+        /**
+         * Checks if a node has `in` operator or not
+         * @param {Node} node The argument node which needs to be checked
+         * @returns {boolean} `true` if there is an `in` operator else `false`.
+         */
+        function isIn(node) {
+            if (
+                node.type === "BinaryExpression" ||
+                node.type === "LogicalExpression"
```

Comment:
I think this wouldn't catch:

```js
/*eslint arrow-body-style: ["error", "as-needed"]*/

for (a = b => { return c in d ? e : f } ;;);

for (a = b => { return c = d in e } ;;);
```

Could we maybe add `"BinaryExpression[operator='in']"(node)` listener and store the info, instead of traversing the node manually?

Reply:
@mdjermanovic here is the output for the cases you mentioned. 

```js
/*eslint arrow-body-style: ["error", "as-needed"]*/

for (a = b => { return c in d ? e : f } ;;);
// output : for (a = b => c in d ? e : f ;;);

for (a = b => { return c = d in e } ;;);
// output : for (a = b => c = d in e ;;);

```

Reply:
> Could we maybe add `"BinaryExpression[operator='in']"(node)` listener and store the info, instead of traversing the node manually?

I added the following

```js
     "BinaryExpression[operator='in']": function(node){
                isIn = false;
            },
      "BinaryExpression[operator='in']:exit": function(node){
                isIn = true;
            }
```
`isIn` is a initialized as `false` inside the `create` method. 
Now in [this line](https://github.com/eslint/eslint/pull/13228/files#diff-c2240f0cdfc25e416753260c2a767fdeR199), checking  
```js
if(isIn){
...
``` 

Is passing all the test cases but not the one you mentioned here https://github.com/eslint/eslint/pull/13228/files#r439331716

```js
/*eslint arrow-body-style: ["error", "as-needed"]*/

for (a = b => { return c in d ? e : f } ;;);
// output : for (a = b => (c in d ? e : f) ;;);

for (a = b => { return c = d in e } ;;);
// output : for (a = b => (c = d in e) ;;);
```

Reply:
We should probably have some sort of stack with one "info" element per arrow function, see for example how it is done in `require-await` rule.

Reply:
@mdjermanovic can you recommend what should be the output for the cases you mentioned?

Like I changed the code to a similar one like `require-await` and the behavior is like this 

```js
// Input 
for (let a = (b, c, d) => { return vb && c in d; }; ;);

// This PR change
for (let a = (b, c, d) => (vb && c in d); ;);

// After similar change to `require-await`
for (let a = (b, c, d) => vb && c in d; ;);
```

Reply:
Output with parens is correct, it's syntax error without parens.

I can't tell what's wrong without the code :) it would be easier if you'd push some version that looks good to you so we could see how to fix those cases that still don't work.